### PR TITLE
[experiment] nkExportDoc node 

### PR DESCRIPTION
**Pull request by krux02**
approach to solve https://github.com/nim-lang/Nim/issues/10046

### CODE REVIEW

File: compiler/ast.nim [hunk 1]
```diff
@@ -223,6 +223,7 @@ type
     nkBreakState,         # special break statement for easier code generation
     nkFuncDef,            # a func
     nkTupleConstr         # a tuple constructor
+    nkExportDoc,          # a node for export marker and doc comment like ``a* ## comment``
```
Comment (timotheecour) [inline compiler/ast.nim]: what if we have a doc comment without a `*`?
I would still want tooling to be able to access doc comment for these.
(or maybe I misunderstood what is meant here?)

File: compiler/msgs.nim [hunk 2]
```diff
@@ -56,6 +56,7 @@ proc newFileInfo(fullPath: AbsoluteFile, projPath: RelativeFile): TFileInfo =
 
 when defined(nimpretty):
   proc fileSection*(conf: ConfigRef; fid: FileIndex; a, b: int): string =
+    assert(fid.int < conf.m.fileInfos.len, "fid: " & $fid.int & " fileInfosLen " & $conf.m.fileInfos.len & " a: " & $a & " b: " & $b)
```
Comment (timotheecour) [inline compiler/msgs.nim]: " fileInfosLen " 
=>
" fileInfosLen: "

File: compiler/ast.nim (hunk 1 above)
Comment (krux02) [reply compiler/ast.nim]: you get a nkExportDoc if you have either a doccomment, an export marker or both.

**Comment by zah** [comment]
I think the parser still has quite a lot of inconsistencies in the handling of keywords. A type such as `var int` will be sometimes parsed as `nkVarTy` and sometimes as `nkCall(ident("var"), ...)` while the new similar `lent` and `sink` modifiers always get the `nkCall` treatment.

Before the AST gets standardized in 0.20, all such inconsistencies should be addressed IMO.

**Comment by krux02** [comment]
@zah while this is true, it is also off topic.

**Comment by Araq** [comment]
> Before the AST gets standardized in 0.20, all such inconsistencies should be addressed IMO.

I don't know how to do that as I don't have any list of these inconsistencies.

**Comment by gemath** [comment]
Will this compile with the new AST node type pretty please?
```nim
import macros

dumpTree:
  # note the asterisk
  customTypeDefMacro myType*:
    typeBody
```
It doesn't now, with the "PostFix" node, and it hinders creating custom exported types with macros very much.

File: compiler/ast.nim (hunk 1 above)
Comment (timotheecour) [reply compiler/ast.nim]:
```suggestion
    nkExportDoc,          # a node for export marker,  doc comment, or both, eg ``a* ## comment``
```

File: compiler/msgs.nim (hunk 2 above)
Comment (timotheecour) [reply compiler/msgs.nim]:
```suggestion
    assert(fid.int < conf.m.fileInfos.len, "fid: " & $fid.int & " fileInfosLen: " & $conf.m.fileInfos.len & " a: " & $a & " b: " & $b)
```

**Comment by Araq** [comment]
> It doesn't now, with the "PostFix" node, and it hinders creating custom exported types with macros very much.

This will remain to be impossible. Sorry.

**Comment by krux02** [comment]
still failing test:
in ccgexpr.binaryExpr: assert(e.sons[1].typ != nil) fails in test tlateboundgenericparams (generics) in statement:
`doAssert t3(100) == "10010"`

**Comment by Araq** [comment]
Failed experiment. Sad, I know.

**Pull request closed without merging**